File: server/internal/services/race_service.go
package services

import (
	"errors"
	"warhammer-api/internal/models"
	"warhammer-api/internal/repositories"
)

type RaceService struct {
	repo repositories.RaceRepository
}

func NewRaceService(repo repositories.RaceRepository) *RaceService {
	return &RaceService{repo}
}

func (s *RaceService) GetAll() ([]models.Race, error) {
	return s.repo.GetAll()
}

func (s *RaceService) GetByID(id int) (*models.Race, error) {
	return s.repo.GetById(id)
}

func (s *RaceService) Create(data *models.Race) (*models.Race, error) {
	if data.Name == "" {
		return nil, errors.New("name cannot be empty")
	}
	return s.repo.Create(data)
}

func (s *RaceService) Update(data *models.Race) (*models.Race, error) {
	return s.repo.Update(data)
}

func (s *RaceService) Delete(id int) error {
	return s.repo.Delete(id)
}

File: server/services/raceService.js
const raceRepository = require('../repositories/raceRepository');

class RaceService {
	async getAll() {
		return await raceRepository.findAll();
	}

	async getById(id) {
		const race = await raceRepository.findById(id);
		if (!race) {
			throw new Error('Race not found');
		}
		return race;
	}

	async create(raceData) {
		return await raceRepository.create(raceData);
	}

	async update(id, raceData) {
		const updated = await raceRepository.update(id, raceData);
		if (!updated) {
			throw new Error('Race not found');
		}
		return updated;
	}

	async delete(id) {
		const deleted = await raceRepository.delete(id);
		if (!deleted) {
			throw new Error('Race not found');
		}
		return deleted;
	}
}

module.exports = new RaceService();